src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php

```diff
@@ -66,4 +94,85 @@ public function getTypeFromFunctionCall(
 		return $defaultReturnType;
 	}
 
+	private function narrowTypeForJsonDecode(FuncCall $funcCall, Scope $scope): Type
+	{
+		$args = $funcCall->getArgs();
+		$isArrayWithoutStdClass = $this->isForceArrayWithoutStdClass($funcCall, $scope);
+
+		$firstArgValue = $args[0]->value;
+		$firstValueType = $scope->getType($firstArgValue);
+
+		if ($firstValueType instanceof ConstantStringType) {
+			return $this->resolveConstantStringType($firstValueType, $isArrayWithoutStdClass);
+		}
+
+		// fallback type
+		if ($isArrayWithoutStdClass === true) {
+			return new MixedType(true, new ObjectType(stdClass::class));
+		}
+
+		return new MixedType(true);
+	}
+
+	/**
+	 * Is "json_decode(..., true)"?
+	 */
+	private function isForceArrayWithoutStdClass(FuncCall $funcCall, Scope $scope): bool
+	{
+		$args = $funcCall->getArgs();
+
+		if (isset($args[1])) {
+			$secondArgValue = $this->resolveMaskValue($args[1]->value, $scope);
```

Comment:
Not sure whether I understand this mask-resolving magic, but maybe it could re-use https://github.com/phpstan/phpstan-src/blob/1.5.x/src/Type/BitwiseFlagHelper.php instead?

Reply:
Yes, please use the new helper instead of ConstExprEvaluator.

Reply:
I've tried it and it works for bit constant. On this line the `true` is also resolved, so we'll need somehow resolve the value.

Reply:
shouldn't it be someting like `if (BooleanType) -handle-bool ELSE -use-bitwiseFlagAnalyser` ?

Reply:
Could you share more code, please? I'm not sure what exactly do you mean.

Reply:
I agree with @staabm, we don't need `constExprEvaluator` here. PHPStan is already a type inference engine by itself.

So `$scope->getType($args[1]->value)` will tell you whether it's `true` or `false`, otherwise you can use the `bitwiseFlagAnalyser`.